### PR TITLE
#166240785 Allow users to  rate articles

### PR DESCRIPTION
**Pull request by nkpremices**
##### What does this PR do?
Have the `ratings` part of the  display articles page working

##### Description of the task to be completed
* Design the page
* Implement the logic
* Write tests

##### Screenshoots
* When an article has already been rated
Screen Shot 2019-09-13 at 10 06 52
* When an article has not been rated
Screen Shot 2019-09-13 at 10 08 11

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -65,15 +65,19 @@
     "prop-types": "^15.7.2",
     "query-string": "^6.8.2",
     "raw-loader": "^3.1.0",
-    "react": "^16.9.0",
-    "react-dom": "^16.9.0",
+    "react": "^16.8.6",
+    "react-css-themr": "^2.1.2",
```
Comment (MCFrank16) [inline package.json]: what does this package do?

File: src/App/App.js [hunk 2]
```diff
@@ -40,6 +41,7 @@ class App extends Component {
             <Route exact path="/login" component={!isAuth ? Login : Home} />
             <Route exact path="/redirect" component={Redirection} />
             <Route exact path="/signup" component={isAuth ? Home : RegistrationComponent} />
+            <Route exact path="/rate" component={Ratings} />
```
Comment (MCFrank16) [inline src/App/App.js]: for a user to rate a particular article he has to go to a certain route @nkpremices I don't understand why? please tell me if I am wrong

File: src/configs/axios.js [hunk 3]
```diff
@@ -1,6 +1,7 @@
 import axios from 'axios';
 
-const baseURL = 'https://authors-heaven.herokuapp.com/';
+// const baseURL = 'https://authors-heaven.herokuapp.com/';
+const baseURL = 'http://localhost:3000';
```
Comment (MCFrank16) [inline src/configs/axios.js]: will you bring back the URL to the previous one from Heroku

**Review by MCFrank16 — APPROVED** [review]

File: src/App/App.js (hunk 2 above)
Comment (nkpremices) [reply src/App/App.js]: I created this route just for design purposes @MCFrank16. Later I'm gonna bind the component to the read article component

File: src/configs/axios.js (hunk 3 above)
Comment (nkpremices) [reply src/configs/axios.js]: Sure @MCFrank16

File: package.json (hunk 1 above)
Comment (nkpremices) [reply package.json]: Oh!! I actually installed them when I was searching for a good package to use for ratings. I'm gonna remove those two

File: src/App/Components/DisplayArticle/Ratings.js [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+      rating: 0,
+      redirect: false,
+      areRatingsShown: false,
+      isLoggedIn: false
+    };
+  }
+
+  componentDidMount() {
+    const { getRatings: getArticleRatings } = this.props;
+    getArticleRatings('What-is-a-Version-1-UUID');
```
Comment (Ramadhan0) [inline src/App/Components/DisplayArticle/Ratings.js]: @nkpremices I think you have to get `slug` from the article you are reading not adding it manually like this.
or maybe if there is a reason behind this, you may explain more about this.

File: src/App/Components/DisplayArticle/Ratings.js [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    if (token) {
+      this.setState({ isLoggedIn: true });
+      const {
+        props: { rate: postRating, getRatings: getArticleRatings },
+        state: { rating: userRating }
+      } = this;
+      if (rating !== userRating) {
+        this.setState({ rating });
+        await postRating(rating, 'What-is-a-Version-1-UUID', token);
+        getArticleRatings('What-is-a-Version-1-UUID');
```
Comment (Ramadhan0) [inline src/App/Components/DisplayArticle/Ratings.js]: same here @nkpremices, if you are using deferent component, you may get `slug` from the parent component

File: src/App/_tests_/ratings.test.js [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+          users: 10,
+          percentage: 27
+        }
+      ]
+    }
+  };
+
+  const rateArticleError = {
+    status: 401,
+    data: { errors: { ratings: 'you cannot rate twice this article, thanks' } }
```
Comment (Ramadhan0) [inline src/App/_tests_/ratings.test.js:104]: @nkpremices is it    ***a user can not rate an `article twice`*** or ***a user can not rate an `article twice with the same rate`***?

File: src/App/Components/DisplayArticle/Ratings.js (hunk 4 above)
Comment (nkpremices) [reply src/App/Components/DisplayArticle/Ratings.js]: This component will be added on @MCFrank16's component.
I will be taking the slug from there

File: src/App/Components/DisplayArticle/Ratings.js (hunk 5 above)
Comment (nkpremices) [reply src/App/Components/DisplayArticle/Ratings.js]: Same here @Ramadhan0

File: src/App/Components/DisplayArticle/Ratings.js (hunk 4 above)
Comment (nkpremices) [reply src/App/Components/DisplayArticle/Ratings.js]: @Ramadhan0

**Comment by nkpremices** [comment]
this PRs is blocked by the `readArticle` PR

File: src/App/Components/DisplayArticle/Rating.js [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+          <div className="rating__fill">
+            <div className={`fill fill--${rating}`} style={{ width: `${percentage}%` }}>
+              _
+            </div>
+          </div>
+        </div>
+        <div
+          className="rating__users"
+          style={{ display: showUsers ? 'block' : 'none' }}
+          // onMouseLeave={this.hide}
```
Comment (MemunaHaruna) [inline src/App/Components/DisplayArticle/Rating.js]: @nkpremices remove commented code

File: src/App/Components/DisplayArticle/Ratings.js [hunk 8]
```diff
@@ -0,0 +1,154 @@
+/* eslint-disable no-restricted-globals */
+/* eslint-disable react/no-unused-state */
```
Comment (MemunaHaruna) [inline src/App/Components/DisplayArticle/Ratings.js]: @nkpremices why do you have an unused state? you should fix the issue, not disable the rule.

File: src/App/Components/DisplayArticle/Ratings.js [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    ratings: propTypes.object,
+    errorMessage: propTypes.string,
+    successMessage: propTypes.string,
+    rate: propTypes.func
+  };
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      average: undefined,
```
Comment (MemunaHaruna) [inline src/App/Components/DisplayArticle/Ratings.js]: @nkpremices any reason why you're using undefined as opposed to null?

check this out: https://medium.com/@arieldi/anyway-stop-setting-your-variables-to-undefined-difference-undefined-vs-null-25f15d907fb4

File: src/App/Components/DisplayArticle/Rating.js (hunk 7 above)
Comment (nkpremices) [reply src/App/Components/DisplayArticle/Rating.js]: Noted @MemunaHaruna

Comment (nkpremices) [reply src/App/Components/DisplayArticle/Rating.js]: removing it

File: src/App/Components/DisplayArticle/Ratings.js (hunk 8 above)
Comment (nkpremices) [reply src/App/Components/DisplayArticle/Ratings.js]: Oh! I just forgot to remove it @MemunaHaruna

Comment (nkpremices) [reply src/App/Components/DisplayArticle/Ratings.js]: lemme do it

File: src/App/_tests_/AvarageCalculator.test.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+        users: 4,
+        percentage: 11
+      },
+      {
+        rating: 4,
+        users: 10,
+        percentage: 27
+      }
+    ]);
+    expect(typeof average).toEqual('number');
```
Comment (MemunaHaruna) [inline src/App/_tests_/AvarageCalculator.test.js]: @nkpremices really? this test ehn

**Review by MemunaHaruna — COMMENTED** [review]
@nkpremices left a few comments

File: src/App/Components/DisplayArticle/Ratings.js (hunk 9 above)
Comment (nkpremices) [reply src/App/Components/DisplayArticle/Ratings.js]: Thank you @MemunaHaruna 
Let me put it to null according to what I've learned from the ressource

File: src/App/Components/DisplayArticle/Rating.js [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    const { props: { areUsersShown, showUsersForRating: showUsers_ } } = this;
+    if (!areUsersShown) {
+      const { state: { showUsers } } = this;
+      this.setState({ showUsers: !showUsers });
+      showUsers_(!areUsersShown);
+    }
+  };
+
+  hide = () => {
+    console.log('hide');
```
Comment (MemunaHaruna) [inline src/App/Components/DisplayArticle/Rating.js]: @nkpremices remove console.log

File: src/App/Components/DisplayArticle/Rating.js [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+          {isEmpty(users) ? (
+            <div className="no-users">
+              No users for this rating
+              {' '}
+              <i className="fas fa-user-times" />
+            </div>
+          ) : (
+            ''
+          )}
+          {pages > 1 ? (
```
Comment (MemunaHaruna) [inline src/App/Components/DisplayArticle/Rating.js:95]: @nkpremices there's a Pagination library in the codebase right now. Have you tried it?

File: src/App/Components/DisplayArticle/Ratings.js (hunk 9 above)
Comment (MemunaHaruna) [reply src/App/Components/DisplayArticle/Ratings.js]: @nkpremices did you forget to change the undefined to null here?

Comment (nkpremices) [reply src/App/Components/DisplayArticle/Ratings.js]: Oh!! I forgot @MemunaHaruna 
Sorry.
Lemme do that

File: src/App/Components/DisplayArticle/Rating.js (hunk 12 above)
Comment (nkpremices) [reply src/App/Components/DisplayArticle/Rating.js:95]: I tried to use it. But it was looking weird on those popups as they are too small @MemunaHaruna

**Pull request merged**